File: src/volume/VolumeBuilder.ts
import type { VolumeData, VolumeDimensions, VolumeSpacing } from '@/types'

export class VolumeBuilder {
  private width: number
  private height: number
  private depth: number
  private spacingX: number = 1.0
  private spacingY: number = 1.0
  private spacingZ: number = 1.0
  private voxelData: Uint16Array
  private built: boolean = false
  private minValue: number = 65535
  private maxValue: number = 0

  constructor(width: number, height: number, depth: number) {
    this.width = width
    this.height = height
    this.depth = depth
    this.voxelData = new Uint16Array(width * height * depth)
  }

  setSpacing(sx: number, sy: number, sz: number): void {
    this.spacingX = sx
    this.spacingY = sy
    this.spacingZ = sz
  }

  addSlice(zIndex: number, sliceData: Uint16Array, sliceSize: number): boolean {
    if (zIndex < 0 || zIndex >= this.depth) {
      return false
    }

    const expectedSize = this.width * this.height
    if (sliceSize !== expectedSize) {
      return false
    }

    const offset = zIndex * this.width * this.height
    this.voxelData.set(sliceData, offset)

    return true
  }

  buildVolume(): boolean {
    this.minValue = 65535
    this.maxValue = 0

    for (let i = 0; i < this.voxelData.length; i++) {
      const value = this.voxelData[i]
      if (value < this.minValue) this.minValue = value
      if (value > this.maxValue) this.maxValue = value
    }

    if (this.minValue > this.maxValue) {
      this.minValue = 0
      this.maxValue = 0
    }

    this.built = true
    return true
  }

  getVoxelData(): Uint16Array {
    return this.voxelData
  }

  getWidth(): number {
    return this.width
  }

  getHeight(): number {
    return this.height
  }

  getDepth(): number {
    return this.depth
  }

  getSpacing(): VolumeSpacing {
    return {
      x: this.spacingX,
      y: this.spacingY,
      z: this.spacingZ
    }
  }

  private isValidCoordinate(x: number, y: number, z: number): boolean {
    return x >= 0 && x < this.width && y >= 0 && y < this.height && z >= 0 && z < this.depth
  }

  getVoxel(x: number, y: number, z: number): number {
    if (!this.isValidCoordinate(x, y, z)) {
      return 0
    }
    const index = z * this.width * this.height + y * this.width + x
    return this.voxelData[index]
  }

  private lerp(a: number, b: number, t: number): number {
    return a + (b - a) * t
  }

  sampleVolume(x: number, y: number, z: number): number {
    let fx = x
    let fy = y
    let fz = z

    if (fx < 0) fx = 0
    if (fx > this.width - 1) fx = this.width - 1
    if (fy < 0) fy = 0
    if (fy > this.height - 1) fy = this.height - 1
    if (fz < 0) fz = 0
    if (fz > this.depth - 1) fz = this.depth - 1

    const x0 = Math.floor(fx)
    const y0 = Math.floor(fy)
    const z0 = Math.floor(fz)
    let x1 = x0 + 1
    let y1 = y0 + 1
    let z1 = z0 + 1

    if (x1 >= this.width) x1 = this.width - 1
    if (y1 >= this.height) y1 = this.height - 1
    if (z1 >= this.depth) z1 = this.depth - 1

    const tx = fx - x0
    const ty = fy - y0
    const tz = fz - z0

    const v000 = this.getVoxel(x0, y0, z0)
    const v100 = this.getVoxel(x1, y0, z0)
    const v010 = this.getVoxel(x0, y1, z0)
    const v110 = this.getVoxel(x1, y1, z0)
    const v001 = this.getVoxel(x0, y0, z1)
    const v101 = this.getVoxel(x1, y0, z1)
    const v011 = this.getVoxel(x0, y1, z1)
    const v111 = this.getVoxel(x1, y1, z1)

    const v00 = this.lerp(v000, v100, tx)
    const v10 = this.lerp(v010, v110, tx)
    const v01 = this.lerp(v001, v101, tx)
    const v11 = this.lerp(v011, v111, tx)

    const v0 = this.lerp(v00, v10, ty)
    const v1 = this.lerp(v01, v11, ty)

    const v = this.lerp(v0, v1, tz)

    return v
  }

  getMinValue(): number {
    return this.minValue
  }

  getMaxValue(): number {
    return this.maxValue
  }

  getVolumeData(): VolumeData {
    if (!this.built) {
      this.buildVolume()
    }

    return {
      dimensions: {
        width: this.width,
        height: this.height,
        depth: this.depth
      } as VolumeDimensions,
      spacing: {
        x: this.spacingX,
        y: this.spacingY,
        z: this.spacingZ
      } as VolumeSpacing,
      voxelData: this.voxelData,
      metadata: null,
      minValue: this.minValue,
      maxValue: this.maxValue
    }
  }
}

export function createVolumeBuilder(width: number, height: number, depth: number): VolumeBuilder {
  return new VolumeBuilder(width, height, depth)
}

export function buildVolumeFromSlices(
  slices: (Uint16Array | Float32Array)[],
  width: number,
  height: number,
  spacing: VolumeSpacing
): VolumeData {
  const depth = slices.length
  const builder = new VolumeBuilder(width, height, depth)
  builder.setSpacing(spacing.x, spacing.y, spacing.z)

  for (let z = 0; z < depth; z++) {
    const sliceData = slices[z]
    if (sliceData instanceof Uint16Array) {
      builder.addSlice(z, sliceData, width * height)
    } else {
      const uint16Data = new Uint16Array(width * height)
      for (let i = 0; i < sliceData.length; i++) {
        uint16Data[i] = Math.max(0, Math.min(65535, Math.round(sliceData[i])))
      }
      builder.addSlice(z, uint16Data, width * height)
    }
  }

  builder.buildVolume()
  return builder.getVolumeData()
}

File: wasm/src/VolumeBuilder.cpp
#include "VolumeBuilder.h"
#include <cstring>
#include <cmath>
#include <algorithm>

VolumeBuilder::VolumeBuilder(int width, int height, int depth)
    : m_width(width)
    , m_height(height)
    , m_depth(depth)
    , m_spacingX(1.0f)
    , m_spacingY(1.0f)
    , m_spacingZ(1.0f)
    , m_built(false)
    , m_minValue(UINT16_MAX)
    , m_maxValue(0)
{
    m_voxelData.resize(static_cast<size_t>(width) * height * depth, 0);
}

VolumeBuilder::~VolumeBuilder() = default;

void VolumeBuilder::setSpacing(float sx, float sy, float sz) {
    m_spacingX = sx;
    m_spacingY = sy;
    m_spacingZ = sz;
}

bool VolumeBuilder::addSlice(int zIndex, const uint16_t* sliceData, int sliceSize) {
    if (zIndex < 0 || zIndex >= m_depth) {
        return false;
    }

    int expectedSize = m_width * m_height;
    if (sliceSize != expectedSize) {
        return false;
    }

    size_t offset = static_cast<size_t>(zIndex) * m_width * m_height;
    std::memcpy(m_voxelData.data() + offset, sliceData, static_cast<size_t>(sliceSize) * sizeof(uint16_t));

    return true;
}

bool VolumeBuilder::buildVolume() {
    m_minValue = UINT16_MAX;
    m_maxValue = 0;

    for (size_t i = 0; i < m_voxelData.size(); ++i) {
        uint16_t value = m_voxelData[i];
        if (value < m_minValue) m_minValue = value;
        if (value > m_maxValue) m_maxValue = value;
    }

    if (m_minValue > m_maxValue) {
        m_minValue = 0;
        m_maxValue = 0;
    }

    m_built = true;
    return true;
}

uint16_t* VolumeBuilder::getVoxelData() const {
    return const_cast<uint16_t*>(m_voxelData.data());
}

bool VolumeBuilder::isValidCoordinate(int x, int y, int z) const {
    return x >= 0 && x < m_width && y >= 0 && y < m_height && z >= 0 && z < m_depth;
}

uint16_t VolumeBuilder::getVoxel(int x, int y, int z) const {
    if (!isValidCoordinate(x, y, z)) {
        return 0;
    }
    size_t index = static_cast<size_t>(z) * m_width * m_height +
                   static_cast<size_t>(y) * m_width +
                   static_cast<size_t>(x);
    return m_voxelData[index];
}

float VolumeBuilder::lerp(float a, float b, float t) {
    return a + (b - a) * t;
}

uint16_t VolumeBuilder::clampVoxelValue(float value) const {
    if (value <= 0.0f) return 0;
    if (value >= static_cast<float>(UINT16_MAX)) return UINT16_MAX;
    return static_cast<uint16_t>(value + 0.5f);
}

float VolumeBuilder::sampleVolume(float x, float y, float z) const {
    float fx = x;
    float fy = y;
    float fz = z;

    if (fx < 0.0f) fx = 0.0f;
    if (fx > static_cast<float>(m_width - 1)) fx = static_cast<float>(m_width - 1);
    if (fy < 0.0f) fy = 0.0f;
    if (fy > static_cast<float>(m_height - 1)) fy = static_cast<float>(m_height - 1);
    if (fz < 0.0f) fz = 0.0f;
    if (fz > static_cast<float>(m_depth - 1)) fz = static_cast<float>(m_depth - 1);

    int x0 = static_cast<int>(std::floor(fx));
    int y0 = static_cast<int>(std::floor(fy));
    int z0 = static_cast<int>(std::floor(fz));
    int x1 = x0 + 1;
    int y1 = y0 + 1;
    int z1 = z0 + 1;

    if (x1 >= m_width) x1 = m_width - 1;
    if (y1 >= m_height) y1 = m_height - 1;
    if (z1 >= m_depth) z1 = m_depth - 1;

    float tx = fx - static_cast<float>(x0);
    float ty = fy - static_cast<float>(y0);
    float tz = fz - static_cast<float>(z0);

    float v000 = static_cast<float>(getVoxel(x0, y0, z0));
    float v100 = static_cast<float>(getVoxel(x1, y0, z0));
    float v010 = static_cast<float>(getVoxel(x0, y1, z0));
    float v110 = static_cast<float>(getVoxel(x1, y1, z0));
    float v001 = static_cast<float>(getVoxel(x0, y0, z1));
    float v101 = static_cast<float>(getVoxel(x1, y0, z1));
    float v011 = static_cast<float>(getVoxel(x0, y1, z1));
    float v111 = static_cast<float>(getVoxel(x1, y1, z1));

    float v00 = lerp(v000, v100, tx);
    float v10 = lerp(v010, v110, tx);
    float v01 = lerp(v001, v101, tx);
    float v11 = lerp(v011, v111, tx);

    float v0 = lerp(v00, v10, ty);
    float v1 = lerp(v01, v11, ty);

    float v = lerp(v0, v1, tz);

    return v;
}
